### PR TITLE
mpc85xx: set correct CPU_TYPE

### PR DESCRIPTION
**Pull request by BKPepe**
This PR contains two PR:

1. It changes CPU_TYPE from 8540 to 8548
2. Adds support for 8548

### CODE REVIEW

**Comment by Ansuel** [comment]
@BKPepe can you rebase?

**Comment by BKPepe** [comment]
I will be at home in 20 minutes then I'd be able to do it.

Dne st 28. 12. 2022 16:28 uživatel Christian Marangi <
***@***.***> napsal:

**Comment by Ansuel** [comment]
sure take your time.

**Comment by neheb** [comment]
I assume this allows better optimization by gcc? That is, I assume it's not like mips24kc being compiled the same as mips34k, 1004k, and some others.

**Comment by BKPepe** [comment]
Not exactly right now. 

The different between ``gcc's mcpu=8540 and -mcpu=8548 flags is support for double precision floating point SPE instructions`` quoted from here https://source.denx.de/u-boot/u-boot/-/commit/c0d0569cf6f7d818cb1bbf5222c5e777bfea4d8c and also there is said ``-mcpu=8548 is just alias to -mcpu=8540.``

Reference for these things can be found here:
- **all the differences between e500v1 and e500v2**
https://www.nxp.com/docs/en/reference-manual/E500CORERM.pdf (page 1-11 or 49) ;
   - Well, I want to introduce SPE instructions later (and reintroduce gcc 8.4.0 back), otherwise the PR would be quite long, so I'm trying to split it as much as I can from [this patch](https://gitlab.nic.cz/turris/os/build/-/raw/hbd/patches/openwrt/to-upstream/0100-mpc85xx-hardware-SPE-float-support.patch).
       - There seems to be [better progress with LLVM](https://reviews.llvm.org/D67787) instead of gcc for PowerPCSPE, but I don't want to get off-topic here.
   
- https://github.com/gcc-mirror/gcc/blob/master/gcc/config/rs6000/vxworks.h#L126

**Comment by BKPepe** [comment]
Should I do another rebase, @ansuel? :)

**Comment by Ansuel** [comment]
Yes pls ahahahhaha

**Comment by BKPepe** [comment]
Aaaand done!

**Comment by jow-** [comment]
@BKPepe @Ansuel - this introduces a new package architecture. If the different CPU type is indeed just an alias for the former one then this change should be reverted. It will require use to build an entirely new package architecture for practically zero gain

**Comment by Ansuel** [comment]
> @BKPepe @Ansuel - this introduces a new package architecture. If the different CPU type is indeed just an alias for the former one then this change should be reverted. It will require us to build an entirely new package architecture for practically zero gain.
> 
> At the very least make sure that the mpc85xx subtargets continue to use the `powerpc_8540` package architecture.

@jow- from Pepe 

The different between gcc's mcpu=8540 and -mcpu=8548 flags is support for double precision floating point SPE instructions quoted from here https://source.denx.de/u-boot/u-boot/-/commit/c0d0569cf6f7d818cb1bbf5222c5e777bfea4d8c and also there is said -mcpu=8548 is just alias to -mcpu=8540.

**Comment by jow-** [comment]
@Ansuel - from the same link:
> Moreover gcc 9 completely removed e500 floating point
support, so since gcc 9 -mcpu=8548 is just alias to -mcpu=8540.

SO I think we shoudl revert that change

**Comment by Ansuel** [comment]
Yep I would wait @bkpepe response and revert

Il Dom 15 Gen 2023, 20:22 Jo-Philipp Wich ***@***.***> ha
scritto:

**Comment by BKPepe** [comment]
Hey @Ansuel, thanks for giving me time to respond! I appreciate it. 

____

The commit _should not_ be reverted because of this reasoning:

1. When it was added here (well, it could be done as part of https://github.com/openwrt/openwrt/commit/0db91251574fd221641a5d89564d244e5406e44f, but as it is missing commit message, it is a different story), it was semi-correct. Thus the valid value for mpc85xx is **8548**, not 8540 as said above. The difference between 8540 and 8548 is, as said, support for double precision floating point SPE, but also, there are other differences as can be seen in the reference manual [^1]. However, the most interesting differences can be seen in  PowerPC™ e500 Core Family Reference Manual [^2].

When can be done as described in this commit: https://github.com/openwrt/openwrt/pull/11629/commits/2cad88b99fdae9766de84e6c1cb56f111eb53748 could be that we force mpc85xx target as CPU_TYPE:=8540, but for the subtargets, we will push the change, but currently, there is **no currently PowerPC e500v1 router supported here.**

2. PowerPC platform, especially the mpc85xx one, is nowadays not very well maintained and it is not produced anymore. 
However, if I am going to compare with mvebu platform (Marvell Armada 385 or Marvell Armada A3720), then the powerpc is a good one if you are going to use Wi-Fi 6 cards, etc., since it works out of the box, there.
   - We got quite interested in mpc85xx these days, mostly @pali, which does recent changes in U-boot and Linux kernel. 
   - We are trying to correct some mistakes, which were done and this is one of it was as the PowerPC CPU_TYPE was set incorrectly...

4. It is a pity that PowerPCSPE got removed in gcc 9 and newer one as there is nobody interested enough to bring it back as it was not maintained. [^3] [^4]

5. Buildroot offers the possibility to have PowerPC SPE support opt-in, and that's the plan to have it also here [^5]. Otherwise, it is using the latest gcc version.

6. The most reason for this pull request was to include this patch, which [introduced SPE float instructions](https://gitlab.nic.cz/turris/os/build/-/merge_requests/542), but we (as Turris) haven't got to submit it to OpenWrt, yet. :-( We confirmed and tested that it works with each OpenWrt release (21.02, 22.03, master) on our Turris 1.x routers. Since those changes are quite long and based on history here, it seems like a good idea to split it into small chunks for better readability and to have it included as long PRs got stalled or nobody has time to review it. 
 - What prevents to send it here is that SPE instructions are forced to use no matter what. It should be opt-in as Buildroot done (see 5. ) to avoid breaking changes, failures, etc. So, it by default, uses the latest gcc as PowerPCSPE is only available in gcc 9 and older (see 4.).
 
 7. Off-topic: Recently, I got told that llvm is working on the PowerPCSPE target, so it would be nice to know if OpenWrt can be compiled with llvm  [^6]
 
 Maybe @autobakterie has something more to add, as I was not directly working on it.

[^1]: https://www.nxp.com/docs/en/application-note/AN2807.pdf (page 2 - heading: ``e500v2 Versus e500v1`` and for some technical guys, could be interested in ``Feature Comparisons`` (same page))
[^2]: https://www.nxp.com/docs/en/reference-manual/E500CORERM.pdf (page 1-11 - heading 1.3.1: ``e500v2 Differences``)
[^3]: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=49854#c7
[^4]: https://www.phoronix.com/news/GCC-Removes-PowerPCSPE
[^5]: https://github.com/beagleboard/buildroot/blob/master/package/gcc/Config.in.host#L17
[^6]: https://reviews.llvm.org/D72014

TL;DR: I try to explain it in https://github.com/openwrt/openwrt/pull/11629#issuecomment-1367155294

**Comment by jow-** [comment]
Since neither using LLVM as compiler nor clinging to gcc 9.x is a viable way forward I really see no point in retaining this no-op change which causes needless resource strain for use.

**Comment by autobakterie** [comment]
A small clarification: PPC SPE support was removed in gcc 9. Therefore it is only available in __gcc 8__ and older, which is the version that needs to be reintroduced should `mpc85xx` be able to utilize the SPE extension.

Regarding benefits of supporting PowerPC SPE: It doesn't end with native double floating point arithmetic instructions. SPE also brings in support for vector integer and single floating point instructions.

**Comment by Ansuel** [comment]
Soooo how to proceed with this?

**Comment by neheb** [comment]
if GCC 8 is not to be reintroduced , revert.

OpenWrt does not support LLVM toolchain.

**Comment by Ansuel** [comment]
@BKPepe I just found time to read the response. To sum up in short
- We currently don't have any router that is 8540
- We have only devices with 8548
- The CPU_TYPE was wrong all along
- 8548 is currently NOT supported by gcc but only by LLVM
- There is currently some work to add 8548 back to gcc with a big patch series

If everything is correct I would move this way:
1. Revert this to restore packages usage
2. Wait for 8548 gcc patch to be merged
3. Resubmit this
4. Drop 8540 packages and use 8548 only

What do you think of this solution? The main complain was that we didn't want to duplicate packages for 2 similar CPU_TYPE but from what i can see we don't support 8540 at all and since 8548 would give so much benefits I would just replace and set that by updating the buildbot.

So yes the idea is to revert this and then reintroduce as soon as we have proper support for this. IMHO it seems strange to change now the CPU_TYPE and under the hood having just an alias. What do you think? Tell me if i get this wrong.

**Comment by autobakterie** [comment]
To somehow illustrate the benefits of utilizing SPE in mpc8548 CPUs, I've compared the CPU times of a simple C program doing multiple rounds of square matrices multiplication. That is by no means close to a real-world performance comparison benchmark, but I think it shows the potential floating point performance gains.

Comparisons of average user-mode CPU times for different number types follow (multiplying 5x5 matrices 10,000,000 times with fixed point and 1,000,000 times with floating point numbers):

```
operand number type | gcc 8 with SPE seconds | gcc 12 seconds | portion of SPE run time
single                     1.01                   5.58            553 %
double                     1.32                   6.07            460 %
signed int 32bit           8.60                   9.69            113 %
signed int 64bit          15.96                  16.33            102 % 
```

Approximately 500 % floating point multiplication performance seems like a significant advantage gained by using SPE instructions produced by a compatible compiler. Is continued gcc 8.4 support for specific targets really such a burden to justify throwing that performance out of the window?

**Comment by autobakterie** [comment]
The basic build was compiled with OpenWrt master toolchain gcc 12.2 configured for MPC85xx target, P2020 subtarget, Default target profile with the following explicitly specified flags:
`-O3 -mcpu=8548 -mtune=8548 -msoft-float`

SPE build was compiled with Turris OS 6 (OpenWrt 21.02) toolchain gcc 8.4 configured for Turris1x with these explicit flags:
`-O3 -mcpu=8548 -mtune=8548 -mabi=spe -mspe -mfloat-gprs=double -msingle-float -mdouble-float -mfpu=none -mhard-float`

I've compared user CPU time as reported by `time` (sys time tends to be 0 for any test variant). Measurement was done on a Turris 1.x router running Turris OS 6.2.2.

4 test program variants differed only in the numeric type used to store the matrices and compute the result. Both floating point versions were served the same input matrices with random values between 0 and 10. Each integer version used its own set of matrices with appropriate random positive or negative values. (see all the inputs linked below)

Each program variant with each build configuration has been run 10 times for each of the 3 corresponding matrices.

The 10 time duration values from each particular _test program variant_ - _input file_ combination were used to compute their mean. The maximum standard variation among all the computed means was 0.017 sec. The final time value for each variant is a mean of the three per-input means (which differed by approximately 0.01 sec at most).

[Here is the C program source.](https://gist.github.com/autobakterie/3a3df5cb4af9d3fc802f09f25dd4b1c0)

Here are the program inputs - each consisting of the desired iteration count, symmetric matrix size and two randomly generated square matrices of proper numeric type:
* https://gist.github.com/autobakterie/4c6fc938a2c8ce6df450f72aab0fb711
* https://gist.github.com/autobakterie/6b1030002d57896bd7ddb0cb2d613669
* https://gist.github.com/autobakterie/add2dcd43be64401a5dad331f63c14d8

**Comment by Ansuel** [comment]
@autobakterie again we can't really consider support an extra old gcc version.... if community care enough we can consider having patch to add support for current gcc version but readding an ancient gcc version is a no go honestly...

**Comment by autobakterie** [comment]
> * There is currently some work to add 8548 back to gcc with a big patch series

What we have done is a patch series allowing building for mpc8548 with SPE IS extension enabled. That works well in Turris OS but is dependent on gcc 8.4. 

As @BKPepe stated - we have a nearly complete rebase of the series for application on OpenWrt master, which, unfortunately, still depends on gcc 8 support.
